custom-pages/gadgets/reports/reports.js

```diff
@@ -1,13 +1,192 @@
+            })
+
+            rec._type = record.getTypeQName()
+            let meta = record.getSystemMetadata()
+            rec["_system.created_on.timestamp"] = meta.created_on.timestamp
+            rec["_system.created_by"] = meta.created_by
+            rec["_system.edited_on.timestamp"] = meta.edited_on.timestamp
+            rec["_system.edited_by"] = meta.edited_by
+
+            //copy all skus' .skuId to top-level prop on rec
+            if (PRODUCT === reportType) {
+                record.skus.forEach(function (sku, index) {
+                    rec[`skus[${index}].skuId`] = sku.skuId
+                })
+            }
+
+            //copy over product page.urlList[0].url
+            if (PAGE === reportType) {
+                if (record.urlList && record.urlList[0]) {
+                    rec.url = record.urlList[0].url
+                }
+            }
+
+            //copy over sku.color[0].hexValue and displayName
+            if (SKU === reportType) {
+                rec.skuId = record.skuId
+
+                if (record.color && record.color[0]) {
+                    rec.hexValue = record.color[0].hexValue
+                    rec.displayName = record.color[0].displayName
+                }
+            }
+
+            //copy over priceSkuList[0].price[ 0-2 ].priceType and priceValue
+            //copy over priceSkuList[0].sku[0].sol, soli, eol, and eoli
+            if (PRICESHEET === reportType) {
+                if (record.priceSkuList && record.priceSkuList[0]) {
+                    record.priceSkuList[0].price.forEach(function (price, index) {
+                        rec[`priceSkuList[0].price[${index}].priceType`] = price.priceType
+                        rec[`priceSkuList[0].price[${index}].priceValue`] = price.priceType
+                        rec[`priceSkuList[0].price[${index}].priceValue`] = price.priceType
+                        rec[`priceSkuList[0].price[${index}].priceValue`] = price.priceType
+                    })
+                }
+            }
+
+            return rec
+        })
+
+        return XLSX.utils.json_to_sheet(nodes)
+    }
+
+    function queryNodesThen(query, callback) {
+        Chain(branch).trap(genericErrorLoggerHalter).queryNodes(query, {limit: -1}).then(callback)
+    }
+
+    function exportHolisticReport() {
+        let workbook = XLSX.utils.book_new()
+
+        Ratchet.block('Generating Report', 'This may take a while...', () => {
+            ReportTypes.forEach((reportType) => {
+                let query = buildQuery(reportType)
+                if (query) {
+                    queryNodesThen(query, function () {
+                        workbook.SheetNames.push(reportType)
+                        workbook.Sheets[reportType] = buildWorksheet({
+                            reportType: reportType,
+                            nodes: this.asArray()
+                        })
+
+                        if (4 === workbook.SheetNames.length) {
```

Comment:
You can check against the size of `ReportTypes`

Reply:
or against the index that you are on

Reply:
They're not guaranteed to finish in order requested so index is not usable. `ReportTypes` is a fixed length as it's just an array of consts

Reply:
Have switched to using `ReportTypes.length`